src/cargo/core/compiler/output_sbom.rs

```diff
@@ -0,0 +1,292 @@
+            rpath: profile.rpath,
+            incremental: profile.incremental,
+            panic: profile.panic.to_string(),
+            rustflags,
+        }
+    }
+}
+
+/// Describes a package dependency
+#[derive(Serialize, Clone, Debug)]
+struct SbomPackage {
+    package_id: PackageIdSpec,
+    package: String,
+    /// Package profile is given when it differs from the root level [`Profile`].
+    #[serde(skip_serializing_if = "Option::is_none")]
+    profile: Option<SbomProfile>,
+    version: Option<Version>,
+    features: Vec<String>,
+    build_type: SbomBuildType,
+    extern_crate_name: String,
+    /// Indices into the `packages` array
+    dependencies: Vec<usize>,
+}
+
+impl SbomPackage {
+    pub fn new(unit_dep: &UnitDep, root_profile: &Profile) -> Self {
+        let package_id = unit_dep.unit.pkg.package_id().to_spec();
+        let package = package_id.name().to_string();
+        let profile = if &unit_dep.unit.profile != root_profile {
+            Some((&unit_dep.unit.profile).into())
+        } else {
+            None
+        };
+        let build_type = if unit_dep.unit.mode.is_run_custom_build() {
+            SbomBuildType::Build
+        } else {
+            SbomBuildType::Normal
+        };
+
+        let version = package_id.version();
+        let features = unit_dep
+            .unit
+            .features
+            .iter()
+            .map(|f| f.to_string())
+            .collect_vec();
+
+        Self {
+            package_id,
+            package,
+            profile,
+            version,
+            features,
+            build_type,
+            extern_crate_name: unit_dep.extern_crate_name.to_string(),
+            dependencies: Vec::new(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct SbomTarget {
+    kind: TargetKind,
+    crate_types: Vec<CrateType>,
+    name: String,
+    edition: String,
+}
+
+impl From<&Target> for SbomTarget {
+    fn from(target: &Target) -> Self {
+        SbomTarget {
+            kind: target.kind().clone(),
+            crate_types: target.kind().rustc_crate_types().clone(),
+            name: target.name().to_string(),
+            edition: target.edition().to_string(),
+        }
+    }
+}
+
+#[derive(Serialize, Clone)]
+struct SbomRustc {
+    version: String,
+    wrapper: Option<PathBuf>,
+    workspace_wrapper: Option<PathBuf>,
+    commit_hash: Option<String>,
+    host: String,
+    verbose_version: String,
+}
+
+impl From<&Rustc> for SbomRustc {
+    fn from(rustc: &Rustc) -> Self {
+        Self {
+            version: rustc.version.to_string(),
+            wrapper: rustc.wrapper.clone(),
+            workspace_wrapper: rustc.workspace_wrapper.clone(),
+            commit_hash: rustc.commit_hash.clone(),
+            host: rustc.host.to_string(),
+            verbose_version: rustc.verbose_version.clone(),
+        }
+    }
+}
+
+#[derive(Serialize)]
+struct Sbom {
+    format_version: SbomFormatVersion<1>,
```

Comment:
Not-blocking: eventually we should move this out into `cargo-util-schemas` and provide `Deserialize` impls so people can use that to read this file.  `SbomFormatVersion` isn't compatible with that.